[OpenACC][CIR] Implement basic lowering for combined constructs

erichkeane opened the pull request:
Combined constructs are emitted a little oddly, in that they are the first ones where there are two operations for a single construct. First, the compute variant is emitted with 'combined(loop)', then the loop operation is emitted with 'combined(<variant>)'. Each gets its own normal terminator.

This patch does not yet implement clauses at all, since that is going to require special attention to make sure we get the emitting of them correct, since certain clauses go to different locations, and need their insertion-points set correctly. So this patch sets it up so that we will emit the 'not implemented' diagnostic for all clauses.

erichkeane committed:
[OpenACC][CIR] Implement basic lowering for combined constructs

Combined constructs are emitted a little oddly, in that they are the
only one where there are two operations for a single construct. First,
the compute variant is emitted with 'combined(loop)', then the loop
operation is emitted with 'combined(<variant>)'. Each gets its own
normal terminator.

This patch does not yet implement clauses at all, since that is going to
require special attention to make sure we get the emitting of them
correct, since certain clauses go to different locations, and need their
insertion-points set correctly. So this patch sets it up so that we will
emit the 'not implemented' diagnostic for all clauses.

razvanlupusoru reviewed (approved):
Thank you!

razvanlupusoru commented:
> in that they are the only one where there are two operations for a single construct.

https://mlir.llvm.org/docs/Dialects/OpenACCDialect/#operation-categories
There are other categories where a single construct leads to multiple operations. See description at the link above about the "second group". I should add combined constructs to that list.

erichkeane commented:
> > in that they are the only one where there are two operations for a single construct.
> 
> https://mlir.llvm.org/docs/Dialects/OpenACCDialect/#operation-categories There are other categories where a single construct leads to multiple operations. See description at the link above about the "second group". I should add combined constructs to that list.

Thanks, corrected!  I meant to more say "that different attributes and operation insertions are going to need to happen", but didn't do a good job of that.  Corrected the commit message

andykaylor reviewed (approved):
lgtm
